### ISSUE TITLE
localised tied scalar referenced elsewhere has strange effects

### ISSUE THREAD

**Issue by p5pRT** [Perl/perl5#11763]
Migrated from [rt.perl.org#104118](https://rt-archive.perl.org/perl5/Ticket/Display.html?id=104118) (status was 'open')

Searchable as RT104118$

**Comment by p5pRT** [comment]
### From @cpansprout

This is based on rt\.cpan\.org \#53064\.

sub TIESCALAR\{bless\[\]\, $\_\[0\]\}
sub FETCH \{ warn fetching; $\_\[0\]\[0\] \}
sub STORE \{ warn "storing $\_\[1\]"; $\_\[0\]\[0\] = $\_\[1\] \}
tie $b\, "";
\*a = \\$b;
$b = "one";
\{
&nbsp;  local $b = "two";
&nbsp;  \{
&nbsp;    my $x = "$a";
\#    my $x = $a;
&nbsp;  \}
\}
warn "\\$b is now $b";

That script prints&#8203;:

storing one at \- line 3\.
fetching at \- line 2\.
storing  at \- line 3\.
storing two at \- line 3\.
storing one at \- line 3\.
fetching at \- line 2\.
$b is now one at \- line 14\.

If you comment out the my $x = "$a" line and uncomment the line below it\, you get&#8203;:

storing one at \- line 3\.
fetching at \- line 2\.
storing  at \- line 3\.
storing two at \- line 3\.
fetching at \- line 2\.
storing two at \- line 3\.
fetching at \- line 2\.
$b is now two at \- line 14\.

So apparently the "$a" is skipping get\-magic\.

In any case\, ‘$b is now one’ seems like the correct output to me\.

The hard part is figuring out how it is supposed to work\.  When you write ‘local $foo’ on a regular scalar\, the \*foo\{SCALAR\} slot is temporarily replaced with another scalar\.  If you have a reference to the original $foo elsewhere\, you can modify it even during the localisation\.  The modified value of $foo is what is visible after the localisation goes out of scope\.

Localised tied scalars are a different creature\.  The scalar is replaced with a new scalar\, like an untied variable\, but then the new scalar is tied to the very same object\.  The old scalar has its private flags copied to the public flags\, so that a simple SvSETMAGIC will restore the previous value\.  But that implementation detail leaks out as through a sieve when a reference to the original scalar exists elsewhere\.  That is why "$a" is skipping FETCH\.

Since the new tied variable is tied to the same object\, conceptually it could be considered the same scalar\, just with a localised \*value\*\.  Most people probably think of it that way\.  If we make the implementation actually work that way \(copy the value to a new scalar and put that on the save stack; assign the value of that temporary scalar back to the tied variable on scope exit\)\, it will at least be self\-consistent\.

Is that a good idea?



---
Flags&#8203;:
&nbsp;    category=core
&nbsp;    severity=low


---
Site configuration information for perl 5\.15\.4&#8203;:

Configured by sprout at Wed Nov  2 09&#8203;:06&#8203;:14 PDT 2011\.

Summary of my perl5 \(revision 5 version 15 subversion 4\) configuration&#8203;:
&nbsp;  Snapshot of&#8203;: f3640611309ab8d6271598d071119f09fd9e8cf0
&nbsp;  Platform&#8203;:
&nbsp;    osname=darwin\, osvers=10\.5\.0\, archname=darwin\-thread\-multi\-2level
&nbsp;    uname='darwin pint\.local 10\.5\.0 darwin kernel version 10\.5\.0&#8203;: fri nov 5 23&#8203;:20&#8203;:39 pdt 2010; root&#8203;:xnu\-1504\.9\.17~1release\_i386 i386 '
&nbsp;    config\_args='\-de \-Doptimize=\-g \-Dusedevel \-Duseithreads \-Dmad'
&nbsp;    hint=recommended\, useposix=true\, d\_sigaction=define
&nbsp;    useithreads=define\, usemultiplicity=define
&nbsp;    useperlio=define\, d\_sfio=undef\, uselargefiles=define\, usesocks=undef
&nbsp;    use64bitint=undef\, use64bitall=undef\, uselongdouble=undef
&nbsp;    usemymalloc=n\, bincompat5005=undef
&nbsp;  Compiler&#8203;:
&nbsp;    cc='cc'\, ccflags ='\-fno\-common \-DPERL\_DARWIN \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'\,
&nbsp;    optimize='\-g'\,
&nbsp;    cppflags='\-fno\-common \-DPERL\_DARWIN \-fno\-strict\-aliasing \-pipe \-fstack\-protector \-I/usr/local/include'
&nbsp;    ccversion=''\, gccversion='4\.2\.1 \(Apple Inc\. build 5664\)'\, gccosandvers=''
&nbsp;    intsize=4\, longsize=4\, ptrsize=4\, doublesize=8\, byteorder=1234
&nbsp;    d\_longlong=define\, longlongsize=8\, d\_longdbl=define\, longdblsize=16
&nbsp;    ivtype='long'\, ivsize=4\, nvtype='double'\, nvsize=8\, Off\_t='off\_t'\, lseeksize=8
&nbsp;    alignbytes=8\, prototype=define
&nbsp;  Linker and Libraries&#8203;:
&nbsp;    ld='env MACOSX\_DEPLOYMENT\_TARGET=10\.3 cc'\, ldflags =' \-fstack\-protector \-L/usr/local/lib'
&nbsp;    libpth=/usr/local/lib /usr/lib
&nbsp;    libs=\-ldbm \-ldl \-lm \-lutil \-lc
&nbsp;    perllibs=\-ldl \-lm \-lutil \-lc
&nbsp;    libc=\, so=dylib\, useshrplib=false\, libperl=libperl\.a
&nbsp;    gnulibc\_version=''
&nbsp;  Dynamic Linking&#8203;:
&nbsp;    dlsrc=dl\_dlopen\.xs\, dlext=bundle\, d\_dlsymun=undef\, ccdlflags=' '
&nbsp;    cccdlflags=' '\, lddlflags=' \-bundle \-undefined dynamic\_lookup \-L/usr/local/lib \-fstack\-protector'

Locally applied patches&#8203;:
&nbsp;    



---
@&#8203;INC for perl 5\.15\.4&#8203;:
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.15\.4/darwin\-thread\-multi\-2level
&nbsp;    /usr/local/lib/perl5/site\_perl/5\.15\.4
&nbsp;    /usr/local/lib/perl5/5\.15\.4/darwin\-thread\-multi\-2level
&nbsp;    /usr/local/lib/perl5/5\.15\.4
&nbsp;    /usr/local/lib/perl5/site\_perl
&nbsp;    \.



---
Environment for perl 5\.15\.4&#8203;:
&nbsp;    DYLD\_LIBRARY\_PATH \(unset\)
&nbsp;    HOME=/Users/sprout
&nbsp;    LANG=en\_US\.UTF\-8
&nbsp;    LANGUAGE \(unset\)
&nbsp;    LD\_LIBRARY\_PATH \(unset\)
&nbsp;    LOGDIR \(unset\)
&nbsp;    PATH=/usr/bin&#8203;:/bin&#8203;:/usr/sbin&#8203;:/sbin&#8203;:/usr/local/bin&#8203;:/usr/X11/bin&#8203;:/usr/local/bin
&nbsp;    PERL\_BADLANG \(unset\)
&nbsp;    SHELL=/bin/bash

**Comment by p5pRT** [comment]
### From @rjbs

\* Father Chrysostomos \<perlbug\-followup@&#8203;perl\.org> \[2011\-11\-20T19&#8203;:37&#8203;:23\]
> The hard part is figuring out how it is supposed to work\.  When you write
> ‘local $foo’ on a regular scalar\, the \*foo\{SCALAR\} slot is temporarily
> replaced with another scalar\.  If you have a reference to the original $foo
> elsewhere\, you can modify it even during the localisation\.  The modified
> value of $foo is what is visible after the localisation goes out of scope\.
> 
> Localised tied scalars are a different creature\.  The scalar is replaced with
> a new scalar\, like an untied variable\, but then the new scalar is tied to the
> very same object\.  The old scalar has its private flags copied to the public
> flags\, so that a simple SvSETMAGIC will restore the previous value\.  But that
> implementation detail leaks out as through a sieve when a reference to the
> original scalar exists elsewhere\.  That is why "$a" is skipping FETCH\.

Not only are they different from untied scalars\, but also from tied arrays and
hashes\.

Tied arrays and hashes cease to be tied after localization\, after some work by
Dave M\. and discussion here&#8203;:

&nbsp;  http&#8203;://www\.xray\.mpe\.mpg\.de/cgi\-bin/w3glimpse2html/perl5\-porters/2010\-04/msg00561\.html

In fact\, tied scalars were also no longer tied after local\, after that thread\,
but then arose another thread&#8203;:

&nbsp;  http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2010\-05/msg00627\.html

David Golden notes that this change broke File&#8203;::chdir and Dave Mitchell says
he'll revert it unless there is some argument\.  There is no contribution to the
thread whatsoever \(other than me noting that I don't really use File&#8203;::chdir
anymore; I'd already switched to the no\-localization File&#8203;::push\)\.  The
scalar\-untying is reverted in 89adf4e8e0de17fedf78328b32920f47635850d0\.

It's a tough nut\.  The current behavior stinks\, and breaking stuff stinks\, and
so does inconsistency between tied scalars and tied arrays\.  If I had my
druthers\, I'd not have the tie survive localization\, but I think this may be
a difficult behavior to sell the deprecation of\.

\-\- 
rjbs

**Comment by p5pRT** [comment]
The RT System itself - Status changed from 'new' to 'open'

**Comment by p5pRT** [comment]
### From @xdg

On Mon\, Sep 17\, 2012 at 6&#8203;:56 PM\, Ricardo Signes
\<perl\.p5p@&#8203;rjbs\.manxome\.org> wrote&#8203;:
> David Golden notes that this change broke File&#8203;::chdir and Dave Mitchell says
> he'll revert it unless there is some argument\.  There is no contribution to the
> thread whatsoever \(other than me noting that I don't really use File&#8203;::chdir
> anymore; I'd already switched to the no\-localization File&#8203;::push\)\.  The
> scalar\-untying is reverted in 89adf4e8e0de17fedf78328b32920f47635850d0\.
>
> It's a tough nut\.  The current behavior stinks\, and breaking stuff stinks\, and
> so does inconsistency between tied scalars and tied arrays\.  If I had my
> druthers\, I'd not have the tie survive localization\, but I think this may be
> a difficult behavior to sell the deprecation of\.

I can only repeat my comments at the time&#8203;:
http&#8203;://www\.xray\.mpe\.mpg\.de/mailing\-lists/perl5\-porters/2010\-05/msg00679\.html

tl;dr&#8203;: I'm happy for it to be consistent even if that breaks stuff\.

More generally\, I find tie\(\) to be a huge PITA and it's on my
fix\-it\-with\-a\-Delorean list shortly after version objects\.

David

\-\- 
David Golden \<xdg@&#8203;xdg\.me>
Take back your inbox\! → http&#8203;://www\.bunchmail\.com/
Twitter/IRC&#8203;: @&#8203;xdg

**Comment by p5pRT** [comment]
### From @jkeenan

On Mon Sep 17 17&#8203;:42&#8203;:54 2012\, xdg@&#8203;xdg\.me wrote&#8203;:

> 
> More generally\, I find tie\(\) to be a huge PITA and it's on my
> fix\-it\-with\-a\-Delorean list shortly after version objects\.
> 


"fix\-it\-with\-a\-Delorean" stumps me \.\.\. as well as DDG\, Google and Bing\!

Example&#8203;:  https://duckduckgo.com/?q=fix-it-with-a-Delorean
\(as of writing\)

**Comment by p5pRT** [comment]
### From @b2gills

On Mon\, Sep 9\, 2013 at 6&#8203;:23 PM\, James E Keenan via RT
\<perlbug\-followup@&#8203;perl\.org> wrote&#8203;:
> On Mon Sep 17 17&#8203;:42&#8203;:54 2012\, xdg@&#8203;xdg\.me wrote&#8203;:
>
>>
>> More generally\, I find tie\(\) to be a huge PITA and it's on my
>> fix\-it\-with\-a\-Delorean list shortly after version objects\.
>>
>
>
> "fix\-it\-with\-a\-Delorean" stumps me \.\.\. as well as DDG\, Google and Bing\!
>
> Example&#8203;:  https://duckduckgo.com/?q=fix-it-with-a-Delorean
> \(as of writing\)
>

"Fix it with a Delorean" is obviously a "Back to the future reference\.

You could also say "fix it with a/the TARDIS"

Basically it means going into the past to prevent the mistake from
ever happening
